apollo-router/src/services/subgraph_service.rs

```diff
@@ -597,104 +593,205 @@ async fn call_http(
+                            .to_graphql_error(None),
+                        )
+                        .build()
+                })
+            })
+        }
+        (ContentType::Other, _, Ok(_bytes)) => {
+            unreachable!(
+                "Should not have gotten here. We should not parse the body of a non json response"
+            )
+        }
+        (_, _, Err(err)) => {
+            // This will add any error that occurred during the fetch to the response including invalid content-type
+            graphql::Response::builder()
+                .error(err.to_graphql_error(None))
+                .build()
+        }
+    };
 
-    if display_body {
-        tracing::info!(
-            http.response.body = %String::from_utf8_lossy(&body), apollo.subgraph.name = %cloned_service_name, "Raw response body from subgraph {cloned_service_name:?} received"
-        );
+    // Add an error for response codes that are not 2xx
+    if !parts.status.is_success() {
+        let status = parts.status;
+        graphql.errors.push(
+            FetchError::SubrequestHttpError {
+                service: service_name.to_string(),
+                status_code: Some(status.as_u16()),
+                reason: format!(
+                    "{}: {}",
+                    status.as_str(),
+                    status.canonical_reason().unwrap_or("Unknown")
+                ),
+            }
+            .to_graphql_error(None),
+        )
     }
 
-    let graphql: graphql::Response =
-        tracing::debug_span!("parse_subgraph_response").in_scope(|| {
-            graphql::Response::from_bytes(&cloned_service_name, body).map_err(|error| {
-                FetchError::SubrequestMalformedResponse {
-                    service: cloned_service_name.clone(),
-                    reason: error.to_string(),
-                }
+    let resp = http::Response::from_parts(parts, graphql);
+    Ok(SubgraphResponse::new_from_response(resp, context))
+}
+
+async fn perform_fetch(
+    mut client: Decompression<Client<HttpsConnector<HttpConnector>>>,
+    service_name: &str,
+    request: Request<Body>,
+) -> Result<(Parts, Result<Bytes, FetchError>), FetchError> {
+    let response = match client.call(request).await {
+        Err(err) => {
+            tracing::error!(fetch_error = format!("{err:?}").as_str());
+            return Err(FetchError::SubrequestHttpError {
+                status_code: None,
+                service: service_name.to_string(),
+                reason: err.to_string(),
+            });
+        }
+        Ok(response) => response,
+    };
+    // Keep our parts, we'll need them later
+    let (parts, body) = response.into_parts();
+
+    // Only try to get the body bytes if the content type was one of the accepted types
+    let body_bytes = match parts.headers.get(header::CONTENT_TYPE).map(|v| v.to_str()) {
+        Some(Ok(content_type))
+            if !content_type.contains(APPLICATION_JSON.essence_str())
```

Comment:
converting the header value to a string and just looking for the presence of the substring is very fragile.

In other parts of the router we use the `mediatype` crate to parse those values:
https://github.com/apollographql/router/blob/dev/apollo-router/src/services/layers/content_negociation.rs#L191-L204

[By spec] (https://www.rfc-editor.org/rfc/rfc9110#field.content-type) the content-type header contains one media type

Reply:
I've basically rewritten everything, so it needs to be reviewed from scratch.